lib/libmbedtls/core/tee_lmd_provider.c

```diff
@@ -2421,72 +2424,263 @@ TEE_Result crypto_mac_final(void *ctx, uint32_t algo, uint8_t *digest,
 }
 
-void crypto_aes_ccm_free_ctx(void *ctx __unused)
+void crypto_aes_ccm_free_ctx(void *ctx)
 {
-	if (ctx)
-		assert(0);
+	struct tee_ccm_state *ccm = ctx;
+
+	if (ccm != NULL) {
+		mbedtls_ccm_free(&ccm->ctx);
+		free(ccm->aad);
+	}
+	free(ccm);
 }
 
-void crypto_aes_ccm_copy_state(void *dst_ctx __unused, void *src_ctx __unused)
+void crypto_aes_ccm_copy_state(void *dst_ctx, void *src_ctx)
 {
+	int lmd_res = 0;
+	struct tee_ccm_state *dst_ccm = dst_ctx;
+	struct tee_ccm_state *src_ccm = src_ctx;
+
+	dst_ccm->aad_len = src_ccm->aad_len;
+	dst_ccm->tag_len = src_ccm->tag_len;
+	dst_ccm->iv_size = src_ccm->iv_size;
+	dst_ccm->payload_len = src_ccm->payload_len;
+	dst_ccm->iv_flag = src_ccm->iv_flag;
+	dst_ccm->aad_totallen = src_ccm->aad_totallen;
+	memcpy(dst_ccm->iv, src_ccm->iv, sizeof(src_ccm->iv));
+	memcpy(dst_ccm->aad, src_ccm->aad, src_ccm->aad_totallen);
+
+	lmd_res = mbedtls_ccm_clone(&dst_ccm->ctx, &src_ccm->ctx);
+	if (lmd_res != 0)
+		FMSG("ccm clone failed, res is 0x%x", -lmd_res);
 }
 
-TEE_Result crypto_aes_ccm_init(void *ctx __unused,
-			       TEE_OperationMode mode __unused,
-			       const uint8_t *key __unused,
-			       size_t key_len __unused,
-			       const uint8_t *nonce __unused,
-			       size_t nonce_len __unused,
-			       size_t tag_len __unused,
-			       size_t aad_len __unused,
-			       size_t payload_len __unused)
+TEE_Result crypto_aes_ccm_init(void *ctx, TEE_OperationMode mode,
+			       const uint8_t *key, size_t key_len,
+			       const uint8_t *nonce, size_t nonce_len,
+			       size_t tag_len, size_t aad_len,
+			       size_t payload_len)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	int lmd_res = 0;
+	struct tee_ccm_state *ccm = ctx;
+	const mbedtls_cipher_info_t *cipher_info = NULL;
+
+	/* Check the key length */
+	if (!key || (key_len > TEE_CCM_KEY_MAX_LENGTH))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* check the nonce */
+	if ((nonce_len > TEE_CCM_NONCE_MAX_LENGTH) ||
+	    (nonce_len > sizeof(ccm->iv))) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+	/* check the tag len */
+	if ((tag_len < 4) || (tag_len > TEE_CCM_TAG_MAX_LENGTH) ||
+	    (tag_len % 2 != 0)) {
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	cipher_info = mbedtls_cipher_info_from_values(MBEDTLS_CIPHER_ID_AES,
+						      key_len * 8,
+						      MBEDTLS_MODE_ECB);
+	if (cipher_info == NULL)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (cipher_info->block_size != 16)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	lmd_res = mbedtls_cipher_setup_info(&ccm->ctx.cipher_ctx, cipher_info);
+	if (lmd_res != 0)
+		return get_tee_result(lmd_res);
+
+	lmd_res = mbedtls_cipher_setkey(&ccm->ctx.cipher_ctx, key,
+					key_len * 8, MBEDTLS_ENCRYPT);
+	if (lmd_res != 0)
+		return get_tee_result(lmd_res);
+
+	ccm->tag_len = tag_len;
+	ccm->payload_len = payload_len;
+	ccm->aad_totallen = aad_len;
+	ccm->iv_size = nonce_len;
+	memcpy(ccm->iv, nonce, nonce_len);
+
+	if ((aad_len != 0) && (ccm->aad == NULL)) {
+		ccm->aad = calloc(1, aad_len);
```

Comment:
Yes, we should check it with mbedTLS team.  I had apply one question for it: https://github.com/ARMmbed/mbedtls/issues/1996
For it now, I think we can add aad_len limitation(0xff00) to avoid it too large.

Reply:
What I was asking for is that the AAD didn't have to be buffered here, instead the data would be consumed as it's supplied. Compare with `crypto_aes_ccm_update_aad()` in `core/lib/libtomcrypt/src/tee_ltc_provider.c`.
However, that doesn't have to be fixed right away, we just need to keep track of it so we don't forget to fix it later.

Reply:
Sorry for misunderstanding it.
As we debugging, we find the `add` can be set only once in mbedTLS(whole add data) which is different with tomcrypt(it can be updated by separate add data). So we add this temp buffer to save the add data.
Of course, if we want to remove it we must let mbedTLS to support it.